File: backend/controllers/adminController.js
const bcrypt = require('bcryptjs');
const { validationResult } = require('express-validator');
const pool = require('../config/db');

/**
 * GET /api/admin/dashboard - Get dashboard statistics
 */
const getDashboard = async (req, res) => {
  try {
    const [userCount] = await pool.query('SELECT COUNT(*) as count FROM users');
    const [storeCount] = await pool.query('SELECT COUNT(*) as count FROM stores');
    const [ratingCount] = await pool.query('SELECT COUNT(*) as count FROM ratings');

    res.json({
      totalUsers: userCount[0].count,
      totalStores: storeCount[0].count,
      totalRatings: ratingCount[0].count,
    });
  } catch (error) {
    console.error('Dashboard error:', error);
    res.status(500).json({ message: 'Internal server error.' });
  }
};

/**
 * POST /api/admin/users - Create a new user (admin can create any role)
 */
const createUser = async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { name, email, password, address, role } = req.body;
    const validRoles = ['admin', 'user', 'store_owner'];
    const userRole = validRoles.includes(role) ? role : 'user';

    // Check if email already exists
    const [existing] = await pool.query('SELECT id FROM users WHERE email = ?', [email]);
    if (existing.length > 0) {
      return res.status(409).json({ message: 'Email already registered.' });
    }

    // Hash password
    const salt = await bcrypt.genSalt(10);
    const passwordHash = await bcrypt.hash(password, salt);

    // Insert user
    const [result] = await pool.query(
      'INSERT INTO users (name, email, password_hash, address, role) VALUES (?, ?, ?, ?, ?)',
      [name, email, passwordHash, address || '', userRole]
    );

    res.status(201).json({ message: 'User created successfully.', userId: result.insertId });
  } catch (error) {
    console.error('Create user error:', error);
    res.status(500).json({ message: 'Internal server error.' });
  }
};

/**
 * GET /api/admin/users - List users with optional filters and sorting
 */
const getUsers = async (req, res) => {
  try {
    const { name, email, address, role, sortBy, sortOrder } = req.query;

    let query = `
      SELECT u.id, u.name, u.email, u.address, u.role, u.created_at,
             CASE WHEN u.role = 'store_owner' THEN (
               SELECT COALESCE(AVG(r.rating), 0) FROM ratings r 
               JOIN stores s ON r.store_id = s.id 
               WHERE s.owner_id = u.id
             ) ELSE NULL END as rating
      FROM users u WHERE 1=1
    `;
    const params = [];

    if (name) {
      query += ' AND u.name LIKE ?';
      params.push(`%${name}%`);
    }
    if (email) {
      query += ' AND u.email LIKE ?';
      params.push(`%${email}%`);
    }
    if (address) {
      query += ' AND u.address LIKE ?';
      params.push(`%${address}%`);
    }
    if (role) {
      query += ' AND u.role = ?';
      params.push(role);
    }

    // Sorting
    const allowedSortFields = ['name', 'email', 'address', 'role', 'created_at'];
    const sortField = allowedSortFields.includes(sortBy) ? sortBy : 'created_at';
    const order = sortOrder === 'asc' ? 'ASC' : 'DESC';
    query += ` ORDER BY u.${sortField} ${order}`;

    const [users] = await pool.query(query, params);
    res.json(users);
  } catch (error) {
    console.error('Get users error:', error);
    res.status(500).json({ message: 'Internal server error.' });
  }
};

/**
 * GET /api/admin/users/:id - Get user details
 */
const getUserById = async (req, res) => {
  try {
    const { id } = req.params;

    const [users] = await pool.query(
      `SELECT u.id, u.name, u.email, u.address, u.role, u.created_at FROM users u WHERE u.id = ?`,
      [id]
    );

    if (users.length === 0) {
      return res.status(404).json({ message: 'User not found.' });
    }

    const user = users[0];

    // If store owner, include their store rating
    if (user.role === 'store_owner') {
      const [stores] = await pool.query(
        `SELECT s.id, s.name, COALESCE(AVG(r.rating), 0) as averageRating
         FROM stores s 
         LEFT JOIN ratings r ON r.store_id = s.id 
         WHERE s.owner_id = ? 
         GROUP BY s.id, s.name`,
        [id]
      );
      user.stores = stores;
      user.rating = stores.length > 0 ? stores[0].averageRating : 0;
    }

    res.json(user);
  } catch (error) {
    console.error('Get user by id error:', error);
    res.status(500).json({ message: 'Internal server error.' });
  }
};

/**
 * POST /api/admin/stores - Create a new store
 */
const createStore = async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { name, email, address, ownerId } = req.body;

    // Check if store email already exists
    const [existing] = await pool.query('SELECT id FROM stores WHERE email = ?', [email]);
    if (existing.length > 0) {
      return res.status(409).json({ message: 'Store email already registered.' });
    }

    // If ownerId provided, verify the user is a store_owner
    if (ownerId) {
      const [owner] = await pool.query('SELECT id, role FROM users WHERE id = ?', [ownerId]);
      if (owner.length === 0) {
        return res.status(404).json({ message: 'Owner user not found.' });
      }
      if (owner[0].role !== 'store_owner') {
        return res.status(400).json({ message: 'Selected user is not a store owner.' });
      }
    }

    const [result] = await pool.query(
      'INSERT INTO stores (name, email, address, owner_id) VALUES (?, ?, ?, ?)',
      [name, email, address || '', ownerId || null]
    );

    res.status(201).json({ message: 'Store created successfully.', storeId: result.insertId });
  } catch (error) {
    console.error('Create store error:', error);
    res.status(500).json({ message: 'Internal server error.' });
  }
};

/**
 * GET /api/admin/stores - List stores with optional filters and sorting
 */
const getStores = async (req, res) => {
  try {
    const { name, email, address, sortBy, sortOrder } = req.query;

    let query = `
      SELECT s.id, s.name, s.email, s.address, s.owner_id,
             COALESCE(AVG(r.rating), 0) as rating,
             COUNT(r.id) as ratingCount
      FROM stores s
      LEFT JOIN ratings r ON r.store_id = s.id
      WHERE 1=1
    `;
    const params = [];

    if (name) {
      query += ' AND s.name LIKE ?';
      params.push(`%${name}%`);
    }
    if (email) {
      query += ' AND s.email LIKE ?';
      params.push(`%${email}%`);
    }
    if (address) {
      query += ' AND s.address LIKE ?';
      params.push(`%${address}%`);
    }

    query += ' GROUP BY s.id, s.name, s.email, s.address, s.owner_id';

    // Sorting
    const allowedSortFields = ['name', 'email', 'address', 'rating'];
    const sortField = allowedSortFields.includes(sortBy) ? sortBy : 'name';
    const order = sortOrder === 'asc' ? 'ASC' : 'DESC';
    query += ` ORDER BY ${sortField === 'rating' ? 'rating' : `s.${sortField}`} ${order}`;

    const [stores] = await pool.query(query, params);
    res.json(stores);
  } catch (error) {
    console.error('Get stores error:', error);
    res.status(500).json({ message: 'Internal server error.' });
  }
};

module.exports = { getDashboard, createUser, getUsers, getUserById, createStore, getStores };

File: backend/src/main/java/com/storerating/api/controller/AdminController.java
package com.storerating.api.controller;

import com.storerating.api.dto.*;
import com.storerating.api.entity.*;
import com.storerating.api.repository.*;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/admin")
public class AdminController {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private StoreRepository storeRepository;

    @Autowired
    private RatingRepository ratingRepository;

    @Autowired
    private ActivityRepository activityRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @GetMapping("/dashboard")
    public ResponseEntity<DashboardStatsResponse> getDashboard() {
        long totalUsers = userRepository.count();
        long totalStores = storeRepository.count();
        long totalRatings = ratingRepository.count();

        return ResponseEntity.ok(DashboardStatsResponse.builder()
                .totalUsers(totalUsers)
                .totalStores(totalStores)
                .totalRatings(totalRatings)
                .build());
    }

    @GetMapping("/activities")
    public ResponseEntity<List<Activity>> getActivities() {
        return ResponseEntity.ok(activityRepository.findAllByOrderByCreatedAtDesc());
    }

    @PostMapping("/users")
    public ResponseEntity<?> createUser(@Valid @RequestBody CreateUserRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Email already registered.");
            return ResponseEntity.status(HttpStatus.CONFLICT).body(response);
        }

        Role role;
        try {
            role = Role.valueOf(request.getRole().toLowerCase());
        } catch (IllegalArgumentException e) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Invalid role specified.");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .address(request.getAddress() != null ? request.getAddress() : "")
                .phone("")
                .location("")
                .storeDescription("")
                .role(role)
                .requestedModerator(false)
                .build();

        User savedUser = userRepository.save(user);

        // Auto-create store if role is store_owner
        if (role == Role.store_owner) {
            Store store = Store.builder()
                    .name(savedUser.getName() + "'s Store")
                    .email(savedUser.getEmail())
                    .address(savedUser.getAddress())
                    .description("")
                    .owner(savedUser)
                    .build();
            storeRepository.save(store);
        }

        // Log activity
        activityRepository.save(Activity.builder()
                .userId(savedUser.getId())
                .userName(savedUser.getName())
                .action("CREATE_USER")
                .details("Admin created user: " + savedUser.getName() + " with role " + savedUser.getRole().name())
                .build());

        Map<String, Object> response = new HashMap<>();
        response.put("message", "User created successfully.");
        response.put("userId", savedUser.getId());
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/users")
    public ResponseEntity<List<AdminUserResponse>> getUsers(
            @RequestParam(required = false) String name,
            @RequestParam(required = false) String email,
            @RequestParam(required = false) String address,
            @RequestParam(required = false) String role,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortOrder) {

        String filterName = (name != null && !name.trim().isEmpty()) ? name.trim() : null;
        String filterEmail = (email != null && !email.trim().isEmpty()) ? email.trim() : null;
        String filterAddress = (address != null && !address.trim().isEmpty()) ? address.trim() : null;

        Role filterRole = null;
        if (role != null && !role.trim().isEmpty()) {
            try {
                filterRole = Role.valueOf(role.trim().toLowerCase());
            } catch (IllegalArgumentException e) {
                // Ignore invalid role
            }
        }

        List<User> users = new ArrayList<>(
                userRepository.findAllFiltered(filterName, filterEmail, filterAddress, filterRole)
        );

        // Sorting
        String finalSortBy = (sortBy != null) ? sortBy : "createdAt";
        String finalSortOrder = (sortOrder != null) ? sortOrder : "desc";
        boolean isAsc = "asc".equalsIgnoreCase(finalSortOrder);

        users.sort((u1, u2) -> {
            int comp;
            switch (finalSortBy.toLowerCase()) {
                case "name":
                    comp = u1.getName().compareToIgnoreCase(u2.getName());
                    break;
                case "email":
                    comp = u1.getEmail().compareToIgnoreCase(u2.getEmail());
                    break;
                case "address":
                    String a1 = u1.getAddress() != null ? u1.getAddress() : "";
                    String a2 = u2.getAddress() != null ? u2.getAddress() : "";
                    comp = a1.compareToIgnoreCase(a2);
                    break;
                case "role":
                    comp = u1.getRole().name().compareToIgnoreCase(u2.getRole().name());
                    break;
                default: // created_at / createdAt
                    comp = u1.getCreatedAt().compareTo(u2.getCreatedAt());
                    break;
            }
            return isAsc ? comp : -comp;
        });

        // Map to response
        List<AdminUserResponse> responseList = users.stream().map(u -> {
            Double avgRating = null;
            if (u.getRole() == Role.store_owner) {
                List<Store> ownedStores = storeRepository.findByOwnerId(u.getId());
                if (!ownedStores.isEmpty()) {
                    List<Long> storeIds = ownedStores.stream().map(Store::getId).collect(Collectors.toList());
                    avgRating = ratingRepository.getAverageRatingForStores(storeIds);
                } else {
                    avgRating = 0.0;
                }
            }

            return AdminUserResponse.builder()
                    .id(u.getId())
                    .name(u.getName())
                    .email(u.getEmail())
                    .address(u.getAddress())
                    .role(u.getRole().name())
                    .createdAt(u.getCreatedAt())
                    .rating(avgRating)
                    .phone(u.getPhone())
                    .location(u.getLocation())
                    .storeDescription(u.getStoreDescription())
                    .requestedModerator(u.getRequestedModerator())
                    .build();
        }).collect(Collectors.toList());

        return ResponseEntity.ok(responseList);
    }

    @GetMapping("/users/{id}")
    public ResponseEntity<?> getUserById(@PathVariable Long id) {
        Optional<User> userOpt = userRepository.findById(id);
        if (userOpt.isEmpty()) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "User not found.");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        }

        User u = userOpt.get();

        Double overallRating = null;
        List<AdminUserDetailResponse.StoreRatingInfo> storesList = new ArrayList<>();

        if (u.getRole() == Role.store_owner) {
            List<Store> ownedStores = storeRepository.findByOwnerId(u.getId());
            if (!ownedStores.isEmpty()) {
                List<Long> storeIds = ownedStores.stream().map(Store::getId).collect(Collectors.toList());
                overallRating = ratingRepository.getAverageRatingForStores(storeIds);

                for (Store s : ownedStores) {
                    Double sAvg = ratingRepository.getAverageRatingForStores(Collections.singletonList(s.getId()));
                    storesList.add(AdminUserDetailResponse.StoreRatingInfo.builder()
                            .id(s.getId())
                            .name(s.getName())
                            .averageRating(sAvg)
                            .build());
                }
            } else {
                overallRating = 0.0;
            }
        }

        AdminUserDetailResponse response = AdminUserDetailResponse.builder()
                .id(u.getId())
                .name(u.getName())
                .email(u.getEmail())
                .address(u.getAddress())
                .role(u.getRole().name())
                .createdAt(u.getCreatedAt())
                .rating(overallRating)
                .stores(storesList)
                .phone(u.getPhone())
                .location(u.getLocation())
                .storeDescription(u.getStoreDescription())
                .requestedModerator(u.getRequestedModerator())
                .build();

        return ResponseEntity.ok(response);
    }

    @PutMapping("/users/{id}/promote-moderator")
    public ResponseEntity<?> promoteModerator(@PathVariable Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (user.getRole() != Role.user && user.getRole() != Role.moderator) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Only Simple Users or existing Moderators can be promoted.");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        user.setRole(Role.moderator);
        user.setRequestedModerator(false);
        userRepository.save(user);

        // Log activity
        activityRepository.save(Activity.builder()
                .action("PROMOTED_TO_MODERATOR")
                .details("Admin promoted user '" + user.getName() + "' to Moderator role.")
                .userId(user.getId())
                .userName(user.getName())
                .build());

        Map<String, String> response = new HashMap<>();
        response.put("message", "User promoted to Moderator successfully.");
        return ResponseEntity.ok(response);
    }

    @PutMapping("/users/{id}/demote-moderator")
    public ResponseEntity<?> demoteModerator(@PathVariable Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (user.getRole() != Role.moderator) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "User is not a Moderator.");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        user.setRole(Role.user);
        user.setRequestedModerator(false);
        userRepository.save(user);

        // Log activity
        activityRepository.save(Activity.builder()
                .action("DEMOTED_FROM_MODERATOR")
                .details("Admin demoted user '" + user.getName() + "' back to Simple User role.")
                .userId(user.getId())
                .userName(user.getName())
                .build());

        Map<String, String> response = new HashMap<>();
        response.put("message", "Moderator demoted to Simple User successfully.");
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/users/{id}")
    public ResponseEntity<?> deleteUser(@PathVariable Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (user.getRole() == Role.admin) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Cannot delete an administrator account.");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // Delete user stores (if store owner)
        if (user.getRole() == Role.store_owner) {
            List<Store> ownedStores = storeRepository.findByOwnerId(user.getId());
            for (Store s : ownedStores) {
                storeRepository.delete(s);
            }
        }

        userRepository.delete(user);

        // Log activity
        activityRepository.save(Activity.builder()
                .action("DELETED_USER")
                .details("Admin deleted user account: " + user.getName() + " (" + user.getEmail() + ")")
                .build());

        Map<String, String> response = new HashMap<>();
        response.put("message", "User account deleted successfully.");
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/reviews/{id}")
    public ResponseEntity<?> deleteReview(@PathVariable Long id) {
        Rating rating = ratingRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Rating not found"));

        ratingRepository.delete(rating);

        // Log activity
        activityRepository.save(Activity.builder()
                .action("DELETED_REVIEW")
                .details("Admin deleted review by " + rating.getUser().getName() + " for store '" + rating.getStore().getName() + "'.")
                .build());

        Map<String, String> response = new HashMap<>();
        response.put("message", "Review deleted successfully.");
        return ResponseEntity.ok(response);
    }

    @PostMapping("/stores")
    public ResponseEntity<?> createStore(@Valid @RequestBody CreateStoreRequest request) {
        if (storeRepository.existsByEmail(request.getEmail())) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Store email already registered.");
            return ResponseEntity.status(HttpStatus.CONFLICT).body(response);
        }

        User owner = null;
        if (request.getOwnerId() != null) {
            Optional<User> ownerOpt = userRepository.findById(request.getOwnerId());
            if (ownerOpt.isEmpty()) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Owner user not found.");
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
            }
            owner = ownerOpt.get();
            if (owner.getRole() != Role.store_owner) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Selected user is not a store owner.");
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
            }
        }

        Store store = Store.builder()
                .name(request.getName())
                .email(request.getEmail())
                .address(request.getAddress() != null ? request.getAddress() : "")
                .description("")
                .owner(owner)
                .build();

        Store savedStore = storeRepository.save(store);

        // Log activity
        activityRepository.save(Activity.builder()
                .action("CREATE_STORE")
                .details("Admin created store: " + savedStore.getName() + " owned by " + (owner != null ? owner.getName() : "None"))
                .build());

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Store created successfully.");
        response.put("storeId", savedStore.getId());
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/stores")
    public ResponseEntity<List<AdminStoreProjection>> getStores(
            @RequestParam(required = false) String name,
            @RequestParam(required = false) String email,
            @RequestParam(required = false) String address,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortOrder) {

        String filterName = (name != null && !name.trim().isEmpty()) ? name.trim() : null;
        String filterEmail = (email != null && !email.trim().isEmpty()) ? email.trim() : null;
        String filterAddress = (address != null && !address.trim().isEmpty()) ? address.trim() : null;

        List<AdminStoreProjection> stores = new ArrayList<>(
                storeRepository.findAllStoresForAdmin(filterName, filterEmail, filterAddress)
        );

        // Sorting
        String finalSortBy = (sortBy != null) ? sortBy : "name";
        String finalSortOrder = (sortOrder != null) ? sortOrder : "desc";
        boolean isAsc = "asc".equalsIgnoreCase(finalSortOrder);

        stores.sort((s1, s2) -> {
            int comp;
            switch (finalSortBy.toLowerCase()) {
                case "rating":
                    comp = Double.compare(s1.getRating(), s2.getRating());
                    break;
                case "email":
                    comp = s1.getEmail().compareToIgnoreCase(s2.getEmail());
                    break;
                case "address":
                    String a1 = s1.getAddress() != null ? s1.getAddress() : "";
                    String a2 = s2.getAddress() != null ? s2.getAddress() : "";
                    comp = a1.compareToIgnoreCase(a2);
                    break;
                default: // name
                    comp = s1.getName().compareToIgnoreCase(s2.getName());
                    break;
            }
            return isAsc ? comp : -comp;
        });

        return ResponseEntity.ok(stores);
    }
}
